frontend/src/metabase/admin/datamodel/components/MetricForm/MetricForm.tsx

```diff
@@ -28,13 +29,17 @@ const QUERY_BUILDER_FEATURES = {
 };
 
 export interface MetricFormProps {
+  disableLeaveConfirmationModal: boolean;
+  route: Route;
   metric?: Metric;
   previewSummary?: string;
   updatePreviewSummary: (previewSummary: string) => void;
   onSubmit: (values: Partial<Metric>) => void;
 }
 
 const MetricForm = ({
+  disableLeaveConfirmationModal,
```

Comment:
Could you move the confirmation modal outside of the form? E.g. something like that

```
useEffect(() => {
  onDirtyChange(dirty);
}, [dirty]);
```

It feels wrong the form is responsible for showing the confirmation modal when the "app" component controls navigation. I think it should be the app component that checks for this flag and shows the modal. By doing that you could remove `disableLeaveConfirmationModal` flag

Reply:
> It feels wrong the form is responsible for showing the confirmation modal when the "app" component controls navigation. 

Agreed :+1:  b6ac7e0c8040acd90809e3bc30c00cb99aa1912e + 2380fb3374ab11149ec80c5a6e3c9726f2c2f70c

It also feels wrong to have `onDirtyChange` - I'd rather lift the entire form state and have a dumb form component.
Refactoring would be quite large (parent component is not in TS, form is) - IMO not for this PR.